backdir/spp/hypermail/hypermail.1
.TH HYPERMAIL 1 "July 31, 1994 - Version 1.02"
.SH NAME
hypermail \- convert mail archives in UNIX box format to HTML pages
.SH SYNOPSIS
.BR "hypermail " [ \-ipux ]
.RB [ \-m
.IR "mailbox" ]
.RB [ \-d
.IR "directory" ]
.RB [ \-l
.IR "label" ]
.RB [ \-a
.IR "URL" ]
.RB [ \-b
.IR "URL" ]
.RB [ \-c
.IR "file" ]
.SH DESCRIPTION
.B hypermail
is a program that takes a file of mail messages in UNIX mailbox format and generates a set of cross-referenced HTML documents. Each file that is created represents a separate message in the mail archive and contains links to other articles, so that the entire archive can be browsed in a number of ways by following links. Archives generated by Hypermail can be incrementally updated, and Hypermail is set by default to only update archives when changes are detected.
.LP
Each HTML file that is generated for a message contains (where applicable): the subject of the article, the name and email address of the sender, the date the article was sent, links to the next and previous messages in the archive, a link to the message the article is in reply to, and a link to the message next in the current thread.
.LP
In addition, Hypermail will convert references in each message to email addresses and URLs to hyperlinks so they can be selected. Email addresses will be converted to 
.B mailto:
URLs. or links to a CGI mail program. Note that not all World-Wide Web browsers support the 
.B mailto:
URL yet. 
.LP
To complement each set of HTML messages, four
.I index files
are created which sort the articles by date received, thread, subject, and author. Each entry in these index files are links to the individual articles and provide a bird's-eye view of every archived message:
.TP
.B date.html
The index of articles sorted by the date they were received by the mail daemon.
.TP
.B thread.html
The index of articles sorted by thread first, then the date they were received.
.TP
.B subject.html
The index of articles sorted by subject. Any
.B "Re:"
prefixes in front of subjects will have been stripped out.
.TP
.B author.html
is the index of articles sorted by the first word of the author's name. If the author's name can't be determined, their email address will be substituted.
.LP
One of the index files will be called
.B index.html
and is the default index that users can go to when entering the archive. In the specified directory, articles will be read out in the order that they were read from a mailbox or standard input. Filenames start at zero and increase
in this fashion:
.B 0000.html, 0001.html, 0002.html,
etc.
.SH OPTIONS
.TP
.B \-i
Reads in articles from standard input.
.TP
.BI \-m " mailbox"
Specifies the mailbox to read articles in from. By default, Hypermail will look for a file called
.B mbox.
Note that the
.B \-i
and
.B \-m
options cannot be used together. Only archives in UNIX mailbox format can be read in - mailboxes of this kind are usually appended RFC 822-compliant articles separated by lines such as "\\nFrom person@site Mon Jan 10 12:34:56 1994".
.TP
.BI \-d " directory"
Specifies the directory to put the HTML files and index files that are created. If the directory doesn't exist, a new one will be created with the name that is specified. If the
.B \-d
option isn't used, Hypermail will look for a directory with the same name as the input mailbox or will create one if needed. 
.TP
.BI \-l " label"
This option tells Hypermail what to call the archive \- the name that is specified will be in the title of the index pages so users know what sort of messages are being archived.
.TP
.BI \-a " URL"
This option includes a link labelled
.B "Other mail archives"
in the index pages to the specified URL. This way users who are looking at the Hypermail archive have the opportunity to go to pointers to other mail archives. By default, this is a link to the parent directory which holds the archive files.
.TP
.BI \-b " URL"
This option includes a link labelled
.B "About this archive"
in the index pages to the specified URL. This way users who are looking at the Hypermail archive have the opportunity to go to information about the archive.
.TP
.BI \-c " file"
This option specifies a configuration file to read settings from. By default, Hypermail will look for a file called
.B .hmrc
in the user's home directory.
.TP
.B \-x
This tells Hypermail to explicitly overwrite any previous HTML files that may exist. Use this option only when it is desirable to completely rewrite the entire archive.
.TP
.B \-u
This updates archives by one message only. With this option, only one email message will be read in from a file or standard input. This message will be added to the end of the existing HTML file archive and will be integrated into it by links and cross-references. All archive index files will be regenerated to include the new message.
.TP
.B \-p
This shows a progress report as Hypermail reads in and writes out messages \- the number of files that Hypermail is reading and writing and the file names of the directory and files created are shown.
.LP
If the
.B -u
option is used and articles are read in from a mailbox file, Hypermail will assume that the file contains only one article, no matter how many files the mailbox may actually contain. Make sure that Hypermail is given only one article when using this option.
.LP
If the mailbox that is being read from is an archive that new messages are always being added to, don't use the
.B \-u
or
.B \-x
options. Hypermail will then read in all the messages given it but will only write new messages that have been appended to the mailbox.
.LP
Note that no matter what options are specified, the index files are always rewritten. The date when Hypermail was last run is included in index pages, so it's easy to tell when the archive was last updated.
.SS
CONFIGURATION OPTIONS
.LP
The following settings can be read in as environment variables or from the specified configuration file. Environment settings are in uppercase. For instance, in the C shell, variables can be set as:
.LP
.PD 0
.RS
.B setenv HM_MBOX "/home/john/my_mailbox"
.LP
.B setenv HM_FILEMODE 0600
.PD
.RE
.LP
In the configuration file, blank lines and lines beginning with a hash mark
.RB ( # )
are ignored. Variables must be in lowercase and separated by values with an equals
.RB ( = )
sign, such as:
.LP
.PD 0
.RS
.B set hm_mbox = "/home/john/my_mailbox"
.LP
.B set hm_filemode = 0600
.PD
.RE
.LP
Settings are read in this order: from the program's hard-wired internal defaults, from environment variables, from command-line options, from the configuration file.
.LP
Below is a list of variables that Hypermail understands. Boolean numbers can have the value of
.B 0
or
.B 1.
.TP
.B HM_CONFIGFILE "filename"
This is the default configuration file to read settings in from. This can only be specified as an environment variable. If the first character is "~", Hypermail will look for the file under the current user's home directory.
.TP
.B HM_MBOX "filename"
This is the default mailbox to read messages in from. Define this with a value of
.B "NONE"
to read from standard input as the default.
.TP
.B HM_ARCHIVES "URL"
This will create a link in the archived index pages labelled
.I "Other mail archives"
to the specified URL. Define as
.B "NONE"
to omit such a link.
.TP
.B HM_ABOUT "URL"
This will create a link in the archived index pages labelled
.I "About this archive"
to the specified URL. Define as
.B "NONE"
to omit such a link.
.TP
.B HM_REVERSE boolean_number
Defining this variable as
.B 1
will reverse-sort the article entries in the date and thread index files by the date they were received. That is, the most recent messages will appear at the top of the index rather than the other way around.
.TP
.B HM_SHOWHEADERS boolean_number
Define this as
.B 1 to show the article header lines in the archived HTML files. These lines typically include the
.B "To:", "From:",
and
.B "Subject:"
information found in most email messages.
.TP
.B HM_SHOWHTML boolean_number
Define as
.B 1
this to show the articles in a proportionally-spaced font rather than a fixed-width (monospace) font. Defining this option will also tell Hypermail to attempt to italicize quoted passages in articles.
.TP
.B HM_LABEL "label name"
Define this as the default label to put in archives. 
.TP
.B HM_DIR "directory"
This is the default directory that Hypermail will look for when creating and updating archives. If defined as
.B "NONE"
the directory will have the same name as the input mailbox.
.TP
.B HM_DIRMODE octal_number
This is an octal number that new directories are set to when they are created.  If the archives will be made publically available, it's a good idea to define this as
.B 0755. 
If files will be updated incrementally with sendmail, this will have to be
.B 0777.
.TP
.B HM_FILEMODE octal_number
This is an octal number that new files are set to when they are created. If the archives will be made publically available, it's a good idea to define this as
.B 0644.
If files will be updated incrementally with sendmail, this will have to be
.B 0666.
.TP
.B HM_OVERWRITE boolean_number
Define as
.B 1
to make Hypermail overwrite existing archives by default. 
.TP
.B HM_INCREMENT boolean_number
Define as
.B 1
to read one article only and append it to existing archives by default.
.TP
.B HM_PROGRESS boolean_number
Define as
.B 1
to always show a progress report as Hypermail works. 
.TP
.B HM_THRDLEVELS number
This specifies the number of thread levels to outline in the thread index. For instance, if
.B THRDLEVELS
is
.B 2,
replies to messages will be indented once in the index, but replies to replies, etc., will only be indented once as well.
.TP
.B HM_DEFAULTINDEX type
This specifies the default index that users can view when entering the archive. Valid types are
.B "date,"
.B "thread,"
.B "author,"
and
.B "subject."
.SH BUGS
.LP
.B Sorting:
In the date and thread index files, note that these lists are sorted by the date the articles were received by the system's mail daemon, not by the date they were written on. The order of articles in the date index may not necessarily match the order in which the article files are written and linked together. Because of this, it is a good idea to make sure the mailbox is sorted by date with the most recent messages towards the bottom.
.LP
RFC 822-compliant headers on multiple lines aren't recognized.
.LP
Forwarded messages with bad headers may be incorrectly handled.
.LP
The URL and email hyperlink parser can only handle one instance of both per line.
.SH AUTHORS
.LP
.B Hypermail
was originally developed and designed by Tom Gruber
.RI ( gruber@ksl.stanford.edu )
for Enterprise Integration Technologies (EIT) in Common Lisp. It was later rewritten in C by Kevin Hughes
.RI ( kevinh@eit.com )
at EIT.
.TP
.B http://www.eit.com/software/hypermail/hypermail.html
The latest documentation can be found at this World-Wide Web address.
.TP
.B http://gummo.stanford.edu/html/hypermail/hypermail.html
Documentation for the Common Lisp version of Hypermail can be found at this address.
.LP
Note that Hypermail is subject to license restrictions. It may be freely used for noncommercial purposes. Unauthorized commercial distribution is prohibited.


backdir/spp/hypermail/hypermail.c
/*
** Copyright (C) 1994, Enterprise Integration Technologies Corp.        
** All Rights Reserved.
** Kevin Hughes, kevinh@eit.com 
** 8/2/94
*/

#define MAIN_FILE
#include "hypermail.h"

int main(argc, argv)
     int argc;
     char **argv;
{
	extern char *optarg;
	extern int optind;
	int i, overwrite, use_stdin, use_mbox, increment;
	char mbox[MAXLINE], label[MAXLINE], dir[MAXLINE],
		archives[MAXLINE], about[MAXLINE], configfile[MAXLINE],
		defaultindex[MAXLINE];

/* Load the hard-wired defaults, then the environment variables.
*/

	strcpy(mbox, MBOX);
	if (getenv("HM_MBOX") != NULL)
		strcpy(mbox, getenv("HM_MBOX"));
	strcpy(archives, ARCHIVES);
	if (getenv("HM_ARCHIVES") != NULL)
		strcpy(archives, getenv("HM_ARCHIVES"));
	strcpy(about, ABOUT);
	if (getenv("HM_ABOUT") != NULL)
		strcpy(about, getenv("HM_ABOUT"));
	strcpy(label, LABEL);
	if (getenv("HM_LABEL") != NULL)
		strcpy(label, getenv("HM_LABEL"));
	strcpy(dir, DIR);
	if (getenv("HM_DIR") != NULL)
		strcpy(dir, getenv("HM_DIR"));
	strcpy(defaultindex, DEFAULTINDEX);
	if (getenv("HM_DEFAULTINDEX") != NULL)
		strcpy(defaultindex, getenv("HM_DEFAULTINDEX"));

	showprogress = PROGRESS;
	if (getenv("HM_PROGRESS") != NULL)
		showprogress = atoi(getenv("HM_PROGRESS"));
	overwrite = OVERWRITE;
	if (getenv("HM_OVERWRITE") != NULL)
		overwrite = atoi(getenv("HM_OVERWRITE"));
	increment = INCREMENT;
	if (getenv("HM_INCREMENT") != NULL)
		increment = atoi(getenv("HM_INCREMENT"));
	reverse = REVERSE;
	if (getenv("HM_REVERSE") != NULL)
		reverse = atoi(getenv("HM_REVERSE"));
	showheaders = SHOWHEADERS;
	if (getenv("HM_SHOWHEADERS") != NULL)
		showheaders = atoi(getenv("HM_SHOWHEADERS"));
	showhtml = SHOWHTML;
	if (getenv("HM_SHOWHTML") != NULL)
		showhtml = atoi(getenv("HM_SHOWHTML"));
	thrdlevels = THRDLEVELS;
	if (getenv("HM_THRDLEVELS") != NULL)
		thrdlevels = atoi(getenv("HM_THRDLEVELS"));
	dirmode = DIRMODE;
	if (getenv("HM_DIRMODE") != NULL)
		dirmode = strtol(getenv("HM_DIRMODE"), (char **) NULL, 0);
	filemode = FILEMODE;
	if (getenv("HM_FILEMODE") != NULL)
		filemode = strtol(getenv("HM_FILEMODE"), (char **) NULL, 0);

	strcpy(configfile, CONFIGFILE);
	if (getenv("HM_CONFIGFILE") != NULL)
		strcpy(configfile, getenv("HM_CONFIGFILE"));

	if (!strcmp(mbox, "NONE"))
		use_stdin = 1;
	else
		use_stdin = 0;

/* ...then use the command-line options...
*/

	use_mbox = 0;
	firstdatenum = lastdatenum = 0;
	while ((i = (int) getopt(argc, argv, "?zhixupm:l:d:a:b:c:")) != -1)
		if ((char) i == 'm') {
			strcpy(mbox, optarg);
			use_mbox = 1;
		}
		else if ((char) i == 'a')
			strcpy(archives, optarg);
		else if ((char) i == 'b')
			strcpy(about, optarg);
		else if ((char) i == 'l')
			strcpy(label, optarg);
		else if ((char) i == 'd')
			strcpy(dir, optarg);
		else if ((char) i == 'c')
			strcpy(configfile, optarg);
		else if ((char) i == 'x')
			overwrite = 1;
		else if ((char) i == 'i')
			use_stdin = 1;
		else if ((char) i == 'p')
			showprogress = 1;
		else if ((char) i == 'u')
			increment = 1;
		else if ((char) i == 'z' || (char) i == 'h' ||
		(char) i == '?')
			usage();
		else
			usage();

/* ...then read the configuration file.
*/

	readconfigs(configfile, mbox, label, dir, archives, about, &overwrite,
		&increment, defaultindex);

/* Default names for directories and labels need to be figured out.
*/

	if (use_stdin && !strcmp(dir, "NONE"))
		strcpy(dir, DIRNAME);
	if (!strcmp(dir, "NONE"))
		strcpy(dir, (strrchr(mbox, '/')) ? (char *) strrchr(mbox, '/')
		+ 1 : mbox);
	if (!strcmp(label, "NONE"))
		strcpy(label, (strrchr(mbox, '/')) ? (char *) strrchr(mbox, '/')
		+ 1 : mbox);

/* Which index file will be called "index.html"?
*/

	strcpy(datename, (!strcmp(defaultindex, "date")) ? INDEXNAME :
	DATENAME);
	strcpy(thrdname, (!strcmp(defaultindex, "thread")) ? INDEXNAME :
	THRDNAME);
	strcpy(subjname, (!strcmp(defaultindex, "subject")) ? INDEXNAME :
	SUBJNAME);
	strcpy(authname, (!strcmp(defaultindex, "author")) ? INDEXNAME :
	AUTHNAME);

	if (use_mbox && use_stdin)
		progerr("Can't read from file and stdin at once.");

	gettimezone();
	getthisyear();
	if (increment) {
		loadoldheaders(dir);
		loadheaders(mbox, use_stdin, 1);
		checkdir(dir);
		writearticles(dir, label, overwrite, bignum);
		fixnextheader(dir, bignum - 1);
#ifdef SHOWREPLIES
		fixreplyheader(dir, bignum);
#endif
		fixthreadheader(dir, bignum);
	}
	else {
		loadheaders(mbox, use_stdin, 0);
		checkdir(dir);
		writearticles(dir, label, overwrite, 0);
	}

/* Always write the index files
*/

	writedates(dir, label, archives, about);
	writethreads(dir, label, archives, about);
	writesubjects(dir, label, archives, about);
	writeauthors(dir, label, archives, about);

	exit(0);
}
